### ISSUE TITLE
Terraform 0.14.11 does not check terraform required_version

### ISSUE THREAD

**Issue by LaurentLesle** [hashicorp/terraform#30105]
### Terraform Version

```
version v0.14.11
```

### Terraform Configuration Files

```terraform

terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 2.80.0"
    }
  }
  required_version = "> 0.15"
  experiments      = [module_variable_optional_attrs]
}
...
```

### Expected Behavior

```terraform
Error: Unsupported Terraform Core version
│ 
│   on main.tf line 9, in terraform:
│    9:   required_version = "> 0.15"
│ 
│ This configuration does not support Terraform version 0.14.11. To proceed, either choose another supported Terraform version or update
│ this version constraint. Version constraints are normally set for good reason, so updating the constraint may lead to other errors or
│ unexpected behavior.
```

### Actual Behavior

Terraform bypass the version check, loads the configuration and fails as the sub-module is using configuration_alias introduced in version 15.

see - https://github.com/Azure/terraform-azurerm-caf-enterprise-scale/blob/main/terraform.tf

```terraform
Error: Variables not allowed

On /home/vscode/.terraform.cache/modules/enterprise_scale/terraform.tf line 9:
Variables may not be used here.


Error: Variables not allowed

On /home/vscode/.terraform.cache/modules/enterprise_scale/terraform.tf line 8:
Variables may not be used here.


Error: Variables not allowed

On /home/vscode/.terraform.cache/modules/enterprise_scale/terraform.tf line 9:
Variables may not be used here.


Error: Variables not allowed

On /home/vscode/.terraform.cache/modules/enterprise_scale/terraform.tf line 8:
Variables may not be used here.


Error: Variables not allowed

On /home/vscode/.terraform.cache/modules/enterprise_scale/terraform.tf line 9:
Variables may not be used here.

➜  caf_eslz git:(2112.int) ✗ t version       
Terraform v0.14.11
+ provider registry.terraform.io/hashicorp/azurerm v2.80.0
+ provider registry.terraform.io/hashicorp/time v0.7.2

Your version of Terraform is out of date! The latest version
is 1.0.11. You can update by downloading from https://www.terraform.io/downloads.html
➜  caf_eslz git:(2112.int) ✗ 
```

**Comment by krowlandson** [comment]
Just to add more context to this, I've done some further testing and it appears that the `required_version` does work in previous versions (including `v0.14.11`) as long as you are using a supported configuration:

![image](https://user-images.githubusercontent.com/12268562/145194658-677b3927-d939-4ca4-a482-cd20adfaad0a.png)

In the example above from @LaurentLesle, the errors being exposed are valid errors when using an unsupported version. However this kind of defeats the point of setting a `required_version` in the first place, as Terraform exposes all the _**known**_ errors that you're trying to supress by specifying supported version(s).

**Comment by alisdair** [comment]
Thanks for reporting this, @LaurentLesle. Which commands are you running when you encounter this issue?

As of #29665, we are checking the `required_version` much earlier now. It's likely that this issue does not occur with the latest Terraform 1.0 release, but I'd like to check.

**Comment by LaurentLesle** [comment]
Just terraform init with a module having configuration_aliases

**Comment by alisdair** [comment]
Thanks! I can confirm that the issue is fixed as of the latest 1.0 patch release.

Reproduced with the following configuration:

```terraform
terraform {
  required_version = "1.1.0"
}

moved {
  from = null_resource.none
  to   = null_resource.some
}

resource "null_resource" "some" {
}
```

Running `terraform init` with 0.15.4:

```shellsession
$ terraform-0.15.4 init
There are some problems with the configuration, described below.

The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
╷
│ Error: Unsupported block type
│
│   on main.tf line 5:
│    5: moved {
│
│ Blocks of type "moved" are not expected here.
╵
```

With 1.0.9 it works as expected:

```terraform
$ terraform-1.0.9 init
╷
│ Error: Unsupported Terraform Core version
│
│   on main.tf line 2, in terraform:
│    2:   required_version = "1.1.0"
│
│ This configuration does not support Terraform version 1.0.9. To proceed,
│ either choose another supported Terraform version or update this version
│ constraint. Version constraints are normally set for good reason, so updating
│ the constraint may lead to other errors or unexpected behavior.
╵
```

We're no longer releasing 0.14 versions, so I'm afraid you'll have to upgrade to at least the 1.0 series to receive this fix. Thanks for the report!
